stan/math/prim/mat/fun/csr_matrix_times_vector.hpp

```diff
@@ -0,0 +1,136 @@
+     * @param m Number of rows in matrix.
+     * @param n Number of columns in matrix.
+     * @param w Vector of non-zero values in matrix.
+     * @param v Column index of each non-zero value, same
+     *          length as w.
+     * @param u Index of where each row starts in w, length equal to
+     *          the number of rows plus one.
+     * @param z Number of non-zero entries in each row of w, length equal to
+     *          the number of rows..
+     * @return Dense vector for the product.
+     * @throw std::domain_error if m and n are not positive or are nan.
+     * @throw std::domain_error if the implied sparse matrix and b are
+     *                          not multiplicable.
+     * @throw std::domain_error if m/n/w/v/u/z are not internally
+     * consistent, as defined by the indexing scheme.  Extractors are
+     * defined in Stan which guarantee a consistent set of m/n/w/v/u/z
+     * for a given sparse matrix.
+     */
+
+    /** \addtogroup csr_format
+      */
+
+    template <typename T1, typename T2>
+    inline
+    Eigen::Matrix<typename boost::math::tools::promote_args<T1, T2>::type, Eigen::Dynamic, 1>
+    csr_matrix_times_vector(const int& m,
+                    const int& n,
+                    const Eigen::Matrix<T1, Eigen::Dynamic, 1>& w,
+                    const std::vector<int>& v,
+                    const std::vector<int>& u,
+                    const std::vector<int>& z,
+                    const Eigen::Matrix<T2, Eigen::Dynamic, 1>& b) {
+      using stan::math::dot_product;
+      using stan::math::check_positive;
+      using stan::math::check_size_match;
+      using stan::math::check_range;
+
+      check_positive("csr_matrix_times_vector", "m", m);
+      check_positive("csr_matrix_times_vector", "n", n);
+      check_size_match("csr_matrix_times_vector", "n", n, "b", b.size());
+      check_size_match("csr_matrix_times_vector", "m", m, "u", u.size()-1);
+      check_size_match("csr_matrix_times_vector", "m", m, "z", z.size());
+      check_size_match("csr_matrix_times_vector", "w", w.size(), "v", v.size());
+      check_size_match("csr_matrix_times_vector", "u/z", u[m-1] + z[m-1]-1, "v", v.size());
+      for (int i=0; i < v.size(); ++i)
+        check_range("csr_matrix_times_vector", "v[]", n, v[i]);
+
+      typedef typename boost::math::tools::promote_args<T1, T2>::type fun_scalar_t;
+      Eigen::Matrix<fun_scalar_t, Eigen::Dynamic, 1>  result(m);
+      result.setZero();
+      for (int row = 0; row < m; ++row) { 
+        int row_end_in_w = (u[row]-stan::error_index::value) + z[row];
+        int i = 0; // index into dot-product segment entries.
+        Eigen::Matrix<fun_scalar_t, Eigen::Dynamic, 1> b_sub(z[row]);
+        b_sub.setZero();
+        for (int nze = u[row]-stan::error_index::value; nze < row_end_in_w; ++nze) {
+          check_range("csr_matrix_times_vector", "j", n, v[nze]);
+          b_sub(i) = b(v[nze]-stan::error_index::value);
```

Comment:
with Rob's Cholesky stuff, we found it was faster to use a `.coefRef()` rather than a `()`. See the subsection on "Basic matrix manipulation" at
http://eigen.tuxfamily.org/dox/group__QuickRefPage.html